File: docs/palindromes-in-peg/scaffold_match_input.py
"""A frozen reversed prefix followed by an incrementally arriving text.

Two global input heads supply the prefix snapshots: before its first letter,
and on its last letter. The reversed pattern needs no copied/reversed array.
Text heads denote cuts and inspect their next cell, so they can wait at the
arrival frontier without moving beyond it or inventing an end marker.
"""
from scaffold_circuit import (Circuit, Value, Ref, NEW, PREVIOUS, TRUE, FALSE,
                              conjunction, disjunction, neg, choose)
from scaffold_circuit_structs import StackPool
from scaffold_circuit_input import InputHead


def input_pools(circuit, operations):
  """Fixed construction counts: each name maps to (append, right, left)."""
  def pool(layout):
    return StackPool(circuit, layout if any(layout.values()) else {"unused": 1})
  heads = {side: pool({name + "." + side: counts[index] for name, counts in operations.items()})
           for side, index in (("l", 1), ("r", 2))}
  units = {name: 3 * (counts[0] + 2 * counts[1]) for name, counts in operations.items()}
  rear = pool({name + ".in." + role: counts[0] for name, counts in operations.items()
               for role in ("B", "B2")})
  reverse = pool({name + ".in.Fr": amount for name, amount in units.items()})
  front = pool({name + ".in.Br": 2 * amount for name, amount in units.items()})
  queues = {"F": front, "WF": front, "Br": front, "B": rear, "WB": rear,
            "B2": rear, "Fr": reverse}
  counters = {}
  for role in ("m", "c"):
    counters[role] = {}
    for side in ("pos", "neg"):
      counts = {name + ".in." + role + "." + side:
                (units[name] if role == "m" else 2 * units[name]) if side == "pos" else
                (units[name] + calls[1] if role == "m" else calls[0] + calls[1])
                for name, calls in operations.items()}
      counters[role][side] = pool(counts)
  return heads, queues, counters


class MatchInput:
  def __init__(self, pools, name):
    self.pattern = InputHead(*pools, name + ".pattern")
    self.text = InputHead(*pools, name + ".text")
    self.circuit = self.pattern.circuit
    self.mode_key = name + ".text_mode"
    self.text_mode = self.circuit.get(PREVIOUS, self.mode_key, (False, True), False).eq(True)

  def start(self, snapshot, enabled=TRUE):
    self.pattern.copy_from(snapshot, enabled)
    self.text.reset(enabled)
    self.text_mode = choose(enabled, neg(snapshot.focus.present()), self.text_mode)

  def append(self, cell, enabled=TRUE):
    self.text.append(cell, enabled)

  def available(self):
    return choose(self.text_mode, self.text.can_right(), self.pattern.focus.present())

  def read(self):
    target = Ref.select(self.text_mode, self.text.peek_right(), self.pattern.focus)
    value = self.circuit.get(target, "input", tuple("ab"), "a")
    return Value.select(target.present(), value, Value.constant(None))

  def move(self, direction, enabled=TRUE):
    if direction not in (-1, 1):
      raise ValueError("unit direction required")
    text_mode = self.text_mode
    if direction == 1:
      self.pattern.left(conjunction(enabled, neg(text_mode)))
      self.text.right(conjunction(enabled, text_mode))
      reached_text = conjunction(enabled, neg(text_mode), neg(self.pattern.focus.present()))
      self.text_mode = disjunction(text_mode, reached_text)
    else:
      at_text_start = neg(self.text.focus.present())
      self.text.left(conjunction(enabled, text_mode, neg(at_text_start)))
      enter_pattern = conjunction(enabled, text_mode, at_text_start)
      self.pattern.right(conjunction(enabled, disjunction(neg(text_mode), at_text_start)))
      self.text_mode = conjunction(text_mode, neg(enter_pattern))

  def copy_from(self, other, enabled=TRUE):
    self.pattern.copy_from(other.pattern, enabled)
    self.text.copy_from(other.text, enabled)
    self.text_mode = choose(enabled, other.text_mode, self.text_mode)

  def finalize(self):
    # push/right already perform the queue's bounded maintenance. Frozen
    # snapshots need no additional rotation work on unrelated input steps.
    self.pattern.finalize(maintain=False)
    self.text.finalize(maintain=False)
    self.circuit.put(self.mode_key, Value.select(self.text_mode, Value.constant(True), Value.constant(False)))


def fixture():
  circuit = Circuit("ab!<>[]xy.")
  operations = {"global.begin": (1, 0, 0), "global.end": (1, 1, 0)}
  for name in ("A", "B"):
    operations[name + ".pattern"] = (0, 1, 1)
    operations[name + ".text"] = (1, 1, 1)
  pools = input_pools(circuit, operations)
  begin, end = [InputHead(*pools, "global." + name) for name in ("begin", "end")]
  a, b = [MatchInput(pools, name) for name in ("A", "B")]
  char = circuit.input()
  arrival = disjunction(char.eq("a"), char.eq("b"))
  circuit.put("input", Value.select(char.eq("b"), Value.constant("b"), Value.constant("a")),
              tuple("ab"), "a")
  active = circuit.get(PREVIOUS, "active", (False, True), False).eq(True)
  for head in (begin, end):
    head.append(NEW, arrival)
  end.right(arrival)
  for head in (a, b):
    head.append(NEW, conjunction(arrival, active))
  a.start(end, char.eq("!"))
  b.start(begin, char.eq("!"))
  active = disjunction(active, char.eq("!"))
  for head, right, left in ((a, ">", "<"), (b, "]", "[")):
    head.move(1, conjunction(active, char.eq(right), head.available()))
    # Left commands are expected to be legal in the test trace. The tape
    # primitives set the fault bit if the pattern's beginning is crossed.
    head.move(-1, conjunction(active, char.eq(left)))
  a.copy_from(b, char.eq("x"))
  b.copy_from(a, char.eq("y"))
  for head in (begin, end):
    head.finalize(maintain=False)
  for name, head in (("A", a), ("B", b)):
    value = head.read()
    circuit.put(name + ".observed", value, (None, "a", "b"), None)
    head.finalize()
  circuit.put("active", Value.select(active, Value.constant(True), Value.constant(False)))
  machine = circuit.machine(conjunction(active, a.available(), b.available(), a.read().equal(b.read())))
  return circuit, (a, b), machine

File: scala/pal/src/main/scala/pal/ScaffoldMatchInput.scala
package pal

import scala.collection.immutable.VectorMap
import Expr.TRUE
import Ref.{NEW, PREVIOUS}
import ScaffoldCircuit.{conjunction, disjunction, neg, choose}
import ScaffoldCircuitInput.{InputHead, HeadPools, CounterPools}

/** A frozen reversed prefix followed by incrementally arriving text.
  * Text heads denote cuts and inspect their next cell, so they can wait at
  * the arrival frontier without moving past it or inventing an end marker.
  */
object ScaffoldMatchInput {
  final case class Pools(heads: HeadPools, queues: HeadPools, counters: CounterPools)

  /** Fixed construction counts: each name maps to (append, right, left). */
  def inputPools(circuit: Circuit, operations: collection.Map[String, (Int, Int, Int)]): Pools = {
    def pool(layout: Iterable[(String, Int)]): StackPool = {
      val entries = layout.toVector
      new StackPool(circuit, if (entries.exists(_._2 != 0)) { entries } else { Vector("unused" -> 1) })
    }
    val heads = VectorMap.from(Vector("l" -> 1, "r" -> 2).map { case (side, index) =>
      side -> pool(operations.map { case (name, counts) =>
        (name + "." + side) -> (if (index == 1) { counts._2 } else { counts._3 })
      })
    })
    val units = VectorMap.from(operations.iterator.map { case (name, counts) => name -> (3 * (counts._1 + 2 * counts._2)) })
    val rear = pool(for ((name, counts) <- operations.toVector; role <- Vector("B", "B2")) yield {
      (name + ".in." + role) -> counts._1
    })
    val reverse = pool(units.map { case (name, amount) => (name + ".in.Fr") -> amount })
    val front = pool(units.map { case (name, amount) => (name + ".in.Br") -> (2 * amount) })
    val queues = VectorMap("F" -> front, "WF" -> front, "Br" -> front, "B" -> rear, "WB" -> rear, "B2" -> rear, "Fr" -> reverse)
    val counters = VectorMap.from(Vector("m", "c").map { role =>
      role -> VectorMap.from(Vector("pos", "neg").map { side =>
        side -> pool(operations.map { case (name, calls) =>
          val amount = if (side == "pos") {
            if (role == "m") { units(name) } else { 2 * units(name) }
          } else if (role == "m") { units(name) + calls._2 } else { calls._1 + calls._2 }
          (name + ".in." + role + "." + side) -> amount
        })
      })
    })
    Pools(heads, queues, counters)
  }

  def fixture(): (Circuit, (MatchInput, MatchInput), Scaffold) = {
    val circuit = new Circuit("ab!<>[]xy.")
    var operations = VectorMap("global.begin" -> (1, 0, 0), "global.end" -> (1, 1, 0))
    for (name <- Vector("A", "B")) {
      operations = operations.updated(name + ".pattern", (0, 1, 1)).updated(name + ".text", (1, 1, 1))
    }
    val pools = inputPools(circuit, operations)
    val begin = new InputHead(pools.heads, pools.queues, pools.counters, "global.begin")
    val end = new InputHead(pools.heads, pools.queues, pools.counters, "global.end")
    val a = new MatchInput(pools, "A")
    val b = new MatchInput(pools, "B")
    val char = circuit.input()
    val arrival = disjunction(char.eqTo('a'), char.eqTo('b'))
    circuit.put("input", Value.select(char.eqTo('b'), Value.constant('b'), Value.constant('a')), Vector('a', 'b'), 'a')
    var active = circuit.get(PREVIOUS, "active", Vector(false, true), false).eqTo(true)
    Vector(begin, end).foreach(_.append(NEW, arrival))
    end.right(arrival)
    Vector(a, b).foreach(_.append(NEW, conjunction(arrival, active)))
    a.start(end, char.eqTo('!'))
    b.start(begin, char.eqTo('!'))
    active = disjunction(active, char.eqTo('!'))
    for ((head, right, left) <- Vector((a, '>', '<'), (b, ']', '['))) {
      head.move(1, conjunction(active, char.eqTo(right), head.available()))
      head.move(-1, conjunction(active, char.eqTo(left)))
    }
    a.copyFrom(b, char.eqTo('x'))
    b.copyFrom(a, char.eqTo('y'))
    Vector(begin, end).foreach(_.commit(maintain = false))
    for ((name, head) <- Vector("A" -> a, "B" -> b)) {
      circuit.put(name + ".observed", head.read(), Vector(None, 'a', 'b'), None)
      head.commit()
    }
    circuit.put("active", Value.select(active, Value.constant(true), Value.constant(false)))
    (circuit, (a, b), circuit.machine(conjunction(active, a.available(), b.available(), a.read().equal(b.read()))))
  }
}

final class MatchInput(pools: ScaffoldMatchInput.Pools, name: String) {
  val pattern = new InputHead(pools.heads, pools.queues, pools.counters, name + ".pattern")
  val text = new InputHead(pools.heads, pools.queues, pools.counters, name + ".text")
  val circuit: Circuit = pattern.circuit
  val modeKey: String = name + ".text_mode"
  var textMode: Expr = circuit.get(PREVIOUS, modeKey, Vector(false, true), false).eqTo(true)

  def start(snapshot: InputHead, enabled: Expr = TRUE): Unit = {
    pattern.copyFrom(snapshot, enabled)
    text.reset(enabled)
    textMode = choose(enabled, neg(snapshot.focus.present()), textMode)
  }
  def append(cell: Ref, enabled: Expr = TRUE): Unit = { text.append(cell, enabled) }
  def available(): Expr = choose(textMode, text.canRight(), pattern.focus.present())
  def read(): Value[Any] = {
    val target = Ref.select(textMode, text.peekRight(), pattern.focus)
    val value = circuit.get(target, "input", Vector('a', 'b'), 'a')
    Value.select(target.present(), value, Value.constant(None))
  }
  def move(direction: Int, enabled: Expr = TRUE): Unit = {
    if (direction != -1 && direction != 1) { throw new IllegalArgumentException("unit direction required") }
    val mode = textMode
    if (direction == 1) {
      pattern.left(conjunction(enabled, neg(mode)))
      text.right(conjunction(enabled, mode))
      val reachedText = conjunction(enabled, neg(mode), neg(pattern.focus.present()))
      textMode = disjunction(mode, reachedText)
    } else {
      val atTextStart = neg(text.focus.present())
      text.left(conjunction(enabled, mode, neg(atTextStart)))
      val enterPattern = conjunction(enabled, mode, atTextStart)
      pattern.right(conjunction(enabled, disjunction(neg(mode), atTextStart)))
      textMode = conjunction(mode, neg(enterPattern))
    }
  }
  def copyFrom(other: MatchInput, enabled: Expr = TRUE): Unit = {
    pattern.copyFrom(other.pattern, enabled)
    text.copyFrom(other.text, enabled)
    textMode = choose(enabled, other.textMode, textMode)
  }
  def commit(): Unit = {
    pattern.commit(maintain = false)
    text.commit(maintain = false)
    circuit.put(modeKey, Value.select(textMode, Value.constant(true), Value.constant(false)))
  }
}
